pybamm/models/base_model.py

```diff
@@ -580,9 +580,8 @@ def build_coupled_variables(self):
                         )
                     )
                     try:
-                        self.variables.update(
-                            submodel.get_coupled_variables(self.variables)
-                        )
+                        submodel_coupled_vars = submodel.get_coupled_variables(self.variables)
+                        self.variables[submodel_name] = submodel_coupled_vars
```

Comment:
variables is already a dict with keys corresponding to variable names and values corresponding to their symbols. you'll need to store the coupled variables that each submodel defines in a new object. 

i don't know if it makes sense to do this when building the model, or if it should be done separately when getting the parameter info. does it make a difference if we store e.g. `self.variables` *and* self.variables_by_submodel` where the latter is a dict of dicts (e.g. {"submodel 1": {"var1": x, ...}, ...}`

Reply:
From what I know, there isn't much benefit to make a separate dict. other from keeping info clean and sorted + avoid searching a large dictionary.

Reply:
> i don't know if it makes sense to do this when building the model, or if it should be done separately when getting the parameter info

I think the latter is better. Accessing elements from a dictionary is usually very fast – even if a dict of dicts is used to print out the information, then it can be "normalised" with the name of a submodel as the key (i.e., print the values of the sub-dictionary, similar to how `pandas` can [normalise JSON](https://pandas.pydata.org/docs/reference/api/pandas.json_normalize.html#pandas-json-normalize)) to account for `by_submodel=True`.

If `by_submodel` is `False`, it can pretty-print the whole dict in a table like it currently does. This method _might_ be better, but would also be harder.

Reply:
Okay.. so what should be done now?

Reply:
Should I add an object in `build_coupled_variables`, e.g. `self.variables_by_submodel` which has a structure:
```
{
    'submodel1': {'coupled_var1': value1, 'coupled_var2': value2, ...},
    'submodel2': {'coupled_var3': value3, 'coupled_var4': value4, ...},
    ...
}
```
The code for this would be the same as before:
```python
```

Reply:
Should I add an object in `build_coupled_variables`, e.g. `self.variables_by_submodel` which has a structure:
```
{
    'submodel1': {'coupled_var1': value1, 'coupled_var2': value2, ...},
    'submodel2': {'coupled_var3': value3, 'coupled_var4': value4, ...},
    ...
}
```
The code for this would be the same as before:
```python
                    try:
                        model_var_copy = self.variables.copy()
                        submodel_coupled_result = submodel.get_coupled_variables(self.variables)
                        self.variables_by_submodel[submodel_name] = {
                            key: submodel_coupled_result[key]
                            for key in submodel_coupled_result
                            if key not in model_var_copy
                        }
                        self.variables.update(
                            submodel.get_coupled_variables(self.variables)
                        )
                        submodels.remove(submodel_name)
```
This stores both `self.variables` and `self.variables_by_submodel`.

Next, We can modify `get_parameter_info` to include `by_submodel=False` too.
In this way, if we set `print_paramter_info(by_submodel=False)`, then `get_parameter_info` will print the usual way.
If we set `print_paramter_info(by_submodel=True)`, then we change the `parameter_info` returned by `get_parameter_info` by submodel wise.


NOTE: I have been coding something for this. I was successful in getting the coupled variables. Now I am experimenting with parameter processing and trying to make the links between `get_parameter_info` and `print_parameter_info` better. Also trying to make the code efficient and cleaner to look at... Will commit soon